Sources/Query.swift

```diff
@@ -1,121 +1,200 @@
+        statement.clause = .UPDATE
+        statement.data = items
+        return self
+    }
+
+    public func insert(items: [String: StatementValueType]) -> Self {
+        statement.clause = .INSERT
+        statement.data = items
+        return self
+    }
+    
+    public func with(key: String, _ op: Operator, _ values: StatementValueType...) -> Self {
+        statement.operation?.append((key, op, values))
+        return self
+    }
+    
+    public func _with(key: String, _ op: Operator, _ values: [StatementValueType]) -> Self {
+        statement.operation?.append((key, op, values))
+        return self
+    }
+    
+    public func andWith(key: String, _ op: Operator, _ values: StatementValueType...) -> Self {
+        statement.operation?.append((key, op, values))
+        statement.andIndexes?.append(statement.operation!.count - 1)
+        return self
+    }
+    
+    public func orWith(key: String, _ op: Operator, _ values: StatementValueType...) -> Self {
+        statement.operation?.append((key, op, values))
+        statement.orIndexes?.append(statement.operation!.count - 1)
+        return self
+    }
+    
+    public func orderBy(key: String, _ order: OrderBy) -> Self {
+        statement.orderBy?.append((key, order))
+        return self
+    }
+    
+    public func groupBy(field: String) -> Self {
+        statement.groupBy = field
+        return self
+    }
+    
+    public func limit(count: Int = 1) -> Self {
+        statement.limit = count
+        return self
+    }
+    
+    public func offset(count: Int = 1) -> Self {
+        statement.offset = count
+        return self
+    }
+    
+    public func list(key: String) -> [String]? {
+        guard let results = Database.driver.execute(statement) else {
+            return nil
+        }
+        
+        var items = [String]()
+        
+        for result in results {
+            for (k, v) in result {
+                if k == key {
+                    items.append(v)
+                }
+            }
+        }
+        
+        return items
+    }
+    
+    public func performQuery(string: String) -> Self {
+        return self
+    }
+    
+/*
+     SELECT role.* FROM user
+     INNER JOIN user_role on user_role.role_id = role.id
+     INNER JOIN role on user_role.user_id = user.id
+*/
+    public func join(table: Model.Type, _ type: Join = .Inner) -> Self? {
+        switch statement.clause {
+        case .SELECT:
+            statement.joins?.append((table.entity, type))
+            return self
+        default:
+            return nil
+        }
+    }
+    
+    public func distinct() -> Self {
+        statement.distinct = true
+        return self
+    }
+
+    // MARK: - Aggregate
+
+    public func count(key: String = "*") -> Int {
+        guard let result = aggregate(.COUNT(key)) else {
+            return -1
+        }
+        return Int(result["COUNT(\(key))"]!) ?? -1
```

Comment:
Individual drivers may call this something else. Is there like a .first?

I'd probably also prefer a `throws` or optional in all these calls vs a magic number